[clang][utils] Add auto mode to reduction script

ilovepi opened the pull request:
The existing script allows us to run creduce/cvise automatically on clang reproducers. However, llvm-reduce is much more efficient for middle or backend issues.

In the early stages, we can just determine which phase the crash happens, and choose the best tool. For now, we only support simple test scripts for opt and llc, but this should cover many more cases.

This feature is behind a new flag, `--auto`, and the default behavior of the script is otherwise unchanged.

Fixes #126827

ilovepi committed:
[clang][utils] Add auto mode to reduction script

The existing script allows us to run creduce/cvise automatically on
clang reproducers. However, llvm-reduce is much more efficient for
middle or backend issues.

In the early stages, we can just determine which phase the crash
happens, and choose the best tool. For now, we only support simple
test scripts for opt and llc, but this should cover many more cases.

This feature is behind a new flag `--auto`, and the default behavior of
the script is otherwise unchanged.

aeubanks reviewed (commented):
thanks for doing this, I really appreciate this. initial quick review, will do more thorough review later

can you link https://github.com/llvm/llvm-project/issues/126827?

ilovepi committed:
Fix up detection for backend failures

ilovepi committed:
Refactor IR extraction and checking

aeubanks commented on clang/utils/reduce-clang-crash.py:
```diff
@@ -424,11 +490,76 @@ def run_creduce(self):
+        print("Found Middle/Backend failure")
+        args = self.clang_args + [
+            "-mllvm",
+            "--print-on-crash",
```
at some point we should also incorporate `llvm/utils/reduce_pipeline.py`, but that's for the future

aeubanks commented on clang/utils/reduce-clang-crash.py:
```diff
@@ -424,11 +490,129 @@ def run_creduce(self):
+            sys.exit("The interestingness test does not pass with '--print-on-crash'.")
+
+        # The output from --print-on-crash has an invalid first line (pass name).
+        remove_first_line(self.ir_file)
```
we should change the extra line to start with `;` to avoid having to do this (I've also been annoyed with this in the past)

aeubanks commented on clang/utils/reduce-clang-crash.py:
```diff
@@ -424,11 +490,76 @@ def run_creduce(self):
+
+        self.opt_level = extract_opt_level(self.clang_args) or "-O2"
+
+        if self.check_expected_output(
```
iiuc `opt` won't crash here because the IR we're passing it has already been through the optimization pipeline. we need to generate the IR with `-Xclang -disable-llvm-passes`

ilovepi replied:
let me double check, but I think opt definitly still crashes here (at least in my test case). it certainly was in my earlier testing (and I'm pretty sure I confirmed it after I refactored this yesterday).

but I'd actually like to get the precise pass from the backtrace and use that to just craft the 

```
opt -passes="function<eager-inv>(drop-unnecessary-assumes,float2int,lower-constant-intrinsics,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O2>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" < crash.ll
```

This is the backtrace I'm seeing locally. I'm not 100% on if any of this is gated on a specific cmake flag we set in a typical build of our toolchain, but having the pass string printed seems like a really easy way to get it to reproduce if we get the IR from -print-on-crash.
```
 ../bin/opt < reduced.ll -O2 -disable-output
opt: /usr/local/google/home/paulkirth/llvm-fork/llvm/include/llvm/Support/Casting.h:109: static bool llvm::isa_impl_cl<llvm::VPWidenCastRecipe, const llvm::VPRecipeBase *>::doit(const From *) [To = llvm::VPWidenCastRecipe, From = const llvm::VPRecipeBase *]: Assertion `Val && "isa<> used on a null pointer"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace and instructions to reproduce the bug.
Stack dump:
0.      Program arguments: ../bin/opt -O2 -disable-output
1.      Running pass "function<eager-inv>(drop-unnecessary-assumes,float2int,lower-constant-intrinsics,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O2>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" on module "<stdin>"
2.      Running pass "loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>" on function "_Z1db"
#0 0x0000555739c1b686 ___interceptor_backtrace ../../../../../../../../../llvm-llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:4556:13
#1 0x0000555739daf2e4 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /usr/local/google/home/paulkirth/llvm-fork/llvm/lib/Support/Unix/Signals.inc:838:8
[1]    3448508 IOT instruction  ../bin/opt -O2 -disable-output < reduced.ll
```

ilovepi replied:
i've added a fallback to just extracting the unmodified IR instead of trying to get it just before the crash w/ `--print-on-crash`. I left the old code in place, but we can probably just remove it, unless we think there is an easy/quick way to use it more thoughtfully.

I tried things w/ a few more cases from the issue tracker, and this seemed to work more reliably/faster.

aeubanks commented on clang/utils/reduce-clang-crash.py:
```diff
@@ -186,22 +237,30 @@ def skip_function(func_name):
 
-    def check_expected_output(self, args=None, filename=None):
+    def check_expected_output(self, cmd=None, args=None, filename=None):
+        if not cmd:
```
having all these default arguments makes it very hard to reason, it would be nice to clean up the callers to explicitly pass things in

ilovepi replied:
ack. I'll look at doing that.

aeubanks reviewed (commented):
we really need to come up with some sort of testing strategy for this, otherwise I think it'll be very hard to maintain this script and make improvements without breaking things (would need some blessed way to crash the compiler, e.g. `#pragma clang __debug crash` crashes the frontend, need to figure out how to crash the middle/backend). I think I'd prefer that we figure that out before making improvements

boomanaiden154 commented:
> would need some blessed way to crash the compiler, e.g. #pragma clang __debug crash crashes the frontend, need to figure out how to crash the middle/backend

What do you think the easiest way to do this is? I'm wondering if we could add clang builtins that trigger verifier errors? Doing that for the middle end would be easy. I'm not sure it would be easy to codegen something that makes it through to the backend though. Introducing IR intrinsics that produce verifier failures is the cleanest I can think of but feels a bit heavyweight.

ilovepi commented on clang/utils/reduce-clang-crash.py:
```diff
@@ -424,11 +490,129 @@ def run_creduce(self):
+        if backend_result == FailureType.BackEnd:
+            return backend_result
+
+        # Try running w/ -emit-llvm to generate an IR file,
```
yeah, I think the strategy is still WIP. Likely this bit will get the most refactoring/has most room for improvement.

ilovepi committed:
Remove duplicated code after refactoring

ilovepi committed:
Don't delete the crash file

ilovepi committed:
Address comments

- remove reliance on default arguments
- restructure IR extraction
- fall back to using clang w/ llvm-reduce
- don't assume all -O options are going to work with other tools

ilovepi committed:
Add tests and update logic to match tool output

Clang and other tools don't have 100% matching output, if we don't check
the prefixes we can use opt or llc, but otherwise the script would
always fall back to using clang.

ilovepi committed:
Feature gate tests to only run if appropriate tools are available
